File: rdkit-sys/wrapper/src/conformer.cc
#include "rust/cxx.h"
#include <GraphMol/Depictor/RDDepictor.h>
#include <GraphMol/DistGeomHelpers/Embedder.h>
#include <GraphMol/GraphMol.h>

namespace RDKit {

int embed_molecule(std::shared_ptr<ROMol> &mol) { return DGeomHelpers::EmbedMolecule(*mol); }

std::unique_ptr<std::vector<int32_t>> embed_multiple_confs(std::shared_ptr<ROMol> &mol, unsigned int num_confs) {
	auto ids = DGeomHelpers::EmbedMultipleConfs(*mol, num_confs);
	return std::make_unique<std::vector<int32_t>>(ids.begin(), ids.end());
}

unsigned int compute_2d_coords(std::shared_ptr<ROMol> &mol) { return RDDepict::compute2DCoords(*mol); }

unsigned int mol_get_num_conformers(const std::shared_ptr<ROMol> &mol) { return mol->getNumConformers(); }

bool conformer_is_3d(const std::shared_ptr<ROMol> &mol, int conf_id) { return mol->getConformer(conf_id).is3D(); }

std::unique_ptr<std::vector<double>> get_atom_pos(const std::shared_ptr<ROMol> &mol, int conf_id,
                                                  unsigned int atom_idx) {
	const auto &pos = mol->getConformer(conf_id).getAtomPos(atom_idx);
	return std::make_unique<std::vector<double>>(std::vector<double>{pos.x, pos.y, pos.z});
}

} // namespace RDKit

File: rdkit-sys/src/bridge/conformer.rs
#[cxx::bridge(namespace = "RDKit")]
pub mod ffi {
    unsafe extern "C++" {
        include!("wrapper/include/conformer.h");

        pub type ROMol = crate::ro_mol_ffi::ROMol;

        pub fn embed_molecule(mol: &mut SharedPtr<ROMol>) -> i32;
        pub fn embed_multiple_confs(
            mol: &mut SharedPtr<ROMol>,
            num_confs: u32,
        ) -> UniquePtr<CxxVector<i32>>;
        pub fn compute_2d_coords(mol: &mut SharedPtr<ROMol>) -> u32;

        pub fn mol_get_num_conformers(mol: &SharedPtr<ROMol>) -> u32;
        pub fn conformer_is_3d(mol: &SharedPtr<ROMol>, conf_id: i32) -> bool;
        pub fn get_atom_pos(
            mol: &SharedPtr<ROMol>,
            conf_id: i32,
            atom_idx: u32,
        ) -> UniquePtr<CxxVector<f64>>;
    }
}
